### DISCUSSION TITLE
static schedule and dynamic schedule of multi-threading in moose

### DISCUSSION THREAD

**Discussion by xiaoruiqi** [idaholab/moose#22987] [Q&A General] [answered]
I've found that the multi-threading used in moose is static scheduled, which becomes rather inefficient when the number of elements is not divisible by the number of threads. All of the additional elements will be assigned to the last thread. Is it possible to turn the static scheduled multi-threading to dynamic scheduled one?  @roystgnr

**Comment by hugary1995** [comment, marked as answer]
I think that imbalance gets smeared out as # elements >> # threads. Dynamic scheduling also comes with its own cost.

**Reply by xiaoruiqi** [reply]
Dear hugary1995, 
I really agree with you, however, when # threads comes close to # elements, this imbalance will make great waste of CPU time, using dynamic scheduling should be a better choice. Could you please show me how to change the static scheduling to dynamic scheduling? I'm not very familiar with the source code at this point.

**Reply by hugary1995** [reply]
Search for `Threads::parallel_reduce` in `FEProblemBase.C`. The ranges are created by libmesh. 

I don't have opinions on how we should do this, and I guess I need to see some profiling results to convince myself that "this imbalance will make greate waste of CPU time", plus I don't think # threads ~= # elements is a typical use case anyways...

Tagging @roystgnr for his opinions on this.

**Reply by hugary1995** [reply]
My understanding is that libmesh isn't set up with dynamic scheduling at all, but again @roystgnr may prove me wrong.

As a separate note, dynamic repartitioning is doable, although I don't think that's what you are after here.

**Reply by xiaoruiqi** [reply]
Thanks for your information, I'll do some research on this problem.

**Reply by GiudGiud** [reply]
If you have so few elements that the number of threads is getting close, then it s likely enough to run in serial.

**Reply by roystgnr** [reply]
Scheduling should depend on your underlying threading model.  We don't do anything fancy at the libMesh level, but IIRC e.g. TBB will split a range into more sub-ranges than you have threads, and will hand off sub-ranges to whichever threads are ready for them as soon as they're ready, so even if your subranges aren't all equally expensive you'll still end up doing a roughly equal amount of work on each processor.  ("roughly" here hides ugly details, of course - make subranges too large and the split doesn't end up being as even, make them too small and overhead increases inversely with size, and I have no idea how TBB decides)

I can't imagine any of this makes any difference when N_threads ≈ N_elem, though.  Unless you're running a subscale model inside each element or something similarly expensive, just talking back and forth within a thread pool probably has enough overhead that you'd be better off not threading at all than trying to divide up work that finely.

All that said, "All of the additional elements will be assigned to the last thread" should almost never happen.  If we have `N` threads and `N*Q+R` (R<N) elements, even the most basic static split ought to be running `Q+1` elements on each of `R` threads and `Q` on each of `N-R`, not `Q+R` on one thread and `Q` on `N-1` threads.  If we can reproduce the former behavior, that's either a bug ... or, ironically, it could be *caused* by dynamic splitting, if the subrange size got chosen exceptionally poorly.

**Reply by xiaoruiqi** [reply]
Yes, you're correct! I'm actually running a multi-scale simulation. And because the simulation inside each element is really computationally expensive, I can clearly observe through `htop` command that `Q+R` elements were handled by one thread.

**Reply by roystgnr** [reply]
Is your libMesh build using TBB or pthreads as the backend?  Could you post a stack trace?
